### PR TITLE
[ctx_profile] Integration test

### PR DESCRIPTION
**Pull request by mtrofin**
Compile with clang a program that's instrumented for contextual profiling and verify a profile can be collected.

### CODE REVIEW

File: compiler-rt/test/ctx_profile/TestCases/generate-context.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  std::cout << Indent << "Guid: " << Node.guid() << std::endl;
+  std::cout << Indent << "Entries: " << Node.entrycount() << std::endl;
+  for (uint32_t I = 0U; I < Node.callsites_size(); ++I)
+    for (const auto *N = Node.subContexts()[I]; N; N = N->next()) {
+      std::cout << Indent << "At Index " << I << ":" << std::endl;
+      printProfile(*N, Indent + Increment, Increment);
+    }
+}
+
+// CHECK: Guid: 11065787667334760794
```
Comment (minglotus-6) [inline compiler-rt/test/ctx_profile/TestCases/generate-context.cpp:50]: A dump drive-by question, I wonder if contextual profile contain block counters? If yes maybe check them as well?

Comment (mtrofin) [reply compiler-rt/test/ctx_profile/TestCases/generate-context.cpp:73]: Thanks! And the test is then more interesting, too. Ptal.

File: compiler-rt/test/ctx_profile/TestCases/generate-context.cpp [hunk 2]
```diff
@@ -20,41 +20,70 @@ extern "C" bool __llvm_ctx_profile_fetch(void *Data,
...
+  }
 }
 }
 
 // Make sure the program actually ran correctly.
 // CHECK: check 1
-// CHECK: check 2
-// CHECK: check 2
+// CHECK: check odd
+// CHECK: check even
```
Comment (minglotus-6) [inline compiler-rt/test/ctx_profile/TestCases/generate-context.cpp:45]: I wonder if there is another `check odd` line (corresponding to the 2nd iteration in the loop)?

Comment (mtrofin) [reply compiler-rt/test/ctx_profile/TestCases/generate-context.cpp]: yup, tightened it.

File: compiler-rt/test/ctx_profile/TestCases/generate-context.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+__attribute__((noinline)) void someFunction(int I) {
+  if (I % 2)
+    printf("check odd\n");
+  else
+    printf("check even\n");
+}
+
+// block inlining because the pre-inliner otherwise will inline this - it's
+// too small.
+__attribute__((noinline)) void the_root() {
```
Comment (snehasish) [inline compiler-rt/test/ctx_profile/TestCases/generate-context.cpp:32]: `theRoot` to be consistent with `someFunction`.

**Review by snehasish — APPROVED** [review]
lgtm

**Comment by dyung** [comment]
Hi @mtrofin, the test you added seems to be failing on a build bot: https://lab.llvm.org/buildbot/#/builders/247/builds/18559

**Comment by mtrofin** [comment]
> Hi @mtrofin, the test you added seems to be failing on a build bot: https://lab.llvm.org/buildbot/#/builders/247/builds/18559
> 
> 
> 
> Can you take a look?

Sorry - was afk. Reverting.

**Comment by mgorny** [comment]
This test fails for me, at least when doing a standalone build:

```
FAIL: CtxProfile-x86_64-linux :: TestCases/generate-context.cpp (2185 of 7045)
******************** TEST 'CtxProfile-x86_64-linux :: TestCases/generate-context.cpp' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 4: mkdir -p /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X8
6_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include
+ mkdir -p /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include
RUN: at line 5: cp /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/llvm/include/llvm/ProfileData/CtxInstrContextNode.h /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include/
+ cp /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/llvm/include/llvm/ProfileData/CtxInstrContextNode.h /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include/
RUN: at line 8: /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/lib/llvm/19/bin/clang  --driver-mode=g++ -ldl -lpthread /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt/test/ctx_profile/TestCases/generate-context.cpp -lclang_rt.ctx_profile -I/var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include -O2 -o /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp.bin -mllvm -profile-context-root=theRoot
+ /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/lib/llvm/19/bin/clang --driver-mode=g++ -ldl -lpthread /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt/test/ctx_profile/TestCases/generate-context.cpp -lclang_rt.ctx_profile -I/var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include -O2 -o /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp.bin -mllvm -profile-context-root=theRoot
x86_64-pc-linux-gnu-ld: error: unable to find library -lclang_rt.ctx_profile
clang: error: linker command failed with exit code 1 (use -v to see invocation)

--

********************
```

The library's at:

```
/var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/lib/clang/19/lib/linux/libclang_rt.ctx_profile-x86_64.a
```

So I guess it's missing the suffix and/or `-L`?

**Comment by mtrofin** [comment]
> This test fails for me, at least when doing a standalone build:
> 
> ```
> FAIL: CtxProfile-x86_64-linux :: TestCases/generate-context.cpp (2185 of 7045)
> ******************** TEST 'CtxProfile-x86_64-linux :: TestCases/generate-context.cpp' FAILED ********************
> Exit Code: 1
> 
> Command Output (stderr):
> --
> RUN: at line 4: mkdir -p /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X8
> 6_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include
> + mkdir -p /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include
> RUN: at line 5: cp /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/llvm/include/llvm/ProfileData/CtxInstrContextNode.h /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include/
> + cp /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/llvm/include/llvm/ProfileData/CtxInstrContextNode.h /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include/
> RUN: at line 8: /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/lib/llvm/19/bin/clang  --driver-mode=g++ -ldl -lpthread /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt/test/ctx_profile/TestCases/generate-context.cpp -lclang_rt.ctx_profile -I/var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include -O2 -o /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp.bin -mllvm -profile-context-root=theRoot
> + /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/lib/llvm/19/bin/clang --driver-mode=g++ -ldl -lpthread /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt/test/ctx_profile/TestCases/generate-context.cpp -lclang_rt.ctx_profile -I/var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp_include -O2 -o /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/test/ctx_profile/X86_64LinuxConfig/TestCases/Output/generate-context.cpp.tmp.bin -mllvm -profile-context-root=theRoot
> x86_64-pc-linux-gnu-ld: error: unable to find library -lclang_rt.ctx_profile
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> 
> --
> 
> ********************
> ```
> 
> The library's at:
> 
> ```
> /var/tmp/portage/sys-libs/compiler-rt-sanitizers-19.0.0_pre20240518/work/compiler-rt_build/lib/clang/19/lib/linux/libclang_rt.ctx_profile-x86_64.a
> ```
> 
> So I guess it's missing the suffix and/or `-L`?

Can you give more context - is this on a bot, or can you give me the cmake invocation - for example, if I do `cmake -GNinja -DCMAKE_BUILD_TYPE=Release ../llvm -DLLVM_ENABLE_PROJECTS="clang;compiler-rt"`, then `ninja check-ctx_profile`, the test passes fine and the library is under `$MY_BUILD_DIR/lib/clang/19/lib/x86_64-unknown-linux-gnu/libclang_rt.ctx_profile.a` - so no suffix. I'm trying to understand what would make it have a suffix, basically.

Thanks!

**Comment by mgorny** [comment]
This is a standalone build against installed LLVM + Clang. Once you install these two, the absolute minimum is roughly:

```
$ cmake -G Ninja <path-to-source>/compiler-rt -DCOMPILER_RT_INCLUDE_TESTS=ON
$ ninja check-ctx_profile 
```

To get tests working, you'd also have to install the libraries onto the system. However, this is enough to reproduce the immediate problem.

**Comment by sscalpone** [comment]
Please test with CMake variable -DCOMPILER_RT_BUILD_SANITIZERS=OFF.

```
CMake Error at cmake/Modules/AddCompilerRT.cmake:355 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:RTSanitizerCommon.x86_64>

  Objects of target "RTSanitizerCommon.x86_64" referenced but no such target
  exists.
Call Stack (most recent call first):
  lib/ctx_profile/CMakeLists.txt:22 (add_compiler_rt_runtime)


CMake Error at cmake/Modules/AddCompilerRT.cmake:355 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:RTSanitizerCommon.x86_64>

  Objects of target "RTSanitizerCommon.x86_64" referenced but no such target
  exists.
Call Stack (most recent call first):
  lib/ctx_profile/CMakeLists.txt:22 (add_compiler_rt_runtime)


CMake Error at cmake/Modules/AddCompilerRT.cmake:355 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:RTSanitizerCommon.x86_64>

  Objects of target "RTSanitizerCommon.x86_64" referenced but no such target
  exists.
Call Stack (most recent call first):
  lib/ctx_profile/CMakeLists.txt:22 (add_compiler_rt_runtime)


CMake Error at cmake/Modules/AddCompilerRT.cmake:355 (add_library):
  No SOURCES given to target: clang_rt.ctx_profile-x86_64
Call Stack (most recent call first):
  lib/ctx_profile/CMakeLists.txt:22 (add_compiler_rt_runtime)

CMake Generate step failed.  Build files cannot be regenerated correctly.
```

**Comment by mtrofin** [comment]
@mgorny on a clean debian bookworm docker image, with cmake 3.25.1, I did this:

(monorepo)
```
mkdir build && cd build
cmake -GNinja -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="clang;compiler-rt" ../llvm
sudo ninja install
```

this installed `/usr/local/lib/clang/19/lib/x86_64-unknown-linux-gnu/libclang_rt.ctx_profile.a` - so still no suffix.

Could you share more details on how you build and install, they appear to be essential to being able to repro? Thanks!

**Comment by dyung** [comment]
> Using -DCOMPILER_RT_BUILD_CTX_PROFILE=OFF seems like the correct solution. That's working for me.
> 
> > Please test with CMake variable -DCOMPILER_RT_BUILD_SANITIZERS=OFF.

We are seeing the same cmake errors in our internal build where we set -DCOMPILER_RT_BUILD_SANITIZERS=OFF. Should we now be setting -DCOMPILER_RT_BUILD_CTX_PROFILE=OFF, or can the change be updated to work correctly when -DCOMPILER_RT_BUILD_SANITIZERS=OFF is defined?

**Comment by mgorny** [comment]
> @mgorny on a clean debian bookworm docker image, with cmake 3.25.1, I did this:

I guess you may need a proper multilib system (`lib`+`lib64`) system for that. Debian uses a custom multiarch, I suppose that could cause this.

> Could you share more details on how you build and install, they appear to be essential to being able to repro? Thanks!

I'm using Gentoo Linux amd64. It's literally `ACCEPT_KEYWORDS=** emerge -1v compiler-rt-sanitizers`.

**Comment by mgorny** [comment]
I'm going to try to find a simpler reproducer.

**Comment by mtrofin** [comment]
> I'm going to try to find a simpler reproducer.

Thanks - I want to make sure I understand what substitutions to add (likely that's what I need to do in the lit.config.py)

But! picking up from @dyung and @sscalpone's comments: if you are already doing `-DCOMPILER_RT_BUILD_SANITIZERS=OFF` and/or `-DCOMPILER_RT_BUILD_MEMPROF=OFF` then the right solution for you should be what @sscalpone suggested, indeed.

**Comment by dyung** [comment]
> > I'm going to try to find a simpler reproducer.
> 
> Thanks - I want to make sure I understand what substitutions to add (likely that's what I need to do in the lit.config.py)
> 
> But! picking up from @dyung and @sscalpone's comments: if you are already doing `-DCOMPILER_RT_BUILD_SANITIZERS=OFF` and/or `-DCOMPILER_RT_BUILD_MEMPROF=OFF` then the right solution for you should be what @sscalpone suggested, indeed.

Thanks for confirming. I can add the define to our build, but would greatly prefer if the build would check whether all of the required components are being built and only then enable it.

**Comment by mgorny** [comment]
> > I'm going to try to find a simpler reproducer.
> 
> Thanks - I want to make sure I understand what substitutions to add (likely that's what I need to do in the lit.config.py)

Ok, could you try if you can reproduce if you do a monorepo build with:

```
cmake -G Ninja <source>/llvm -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLLVM_CCACHE_BUILD=ON -DLLVM_TARGETS_TO_BUILD=host -DLLVM_ENABLE_PROJECTS='clang;compiler-rt' -DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF
```

(`-DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF` should be the key option, the rest is just my "quick build" defaults)

**Comment by mtrofin** [comment]
> > > I'm going to try to find a simpler reproducer.
> > 
> > 
> > Thanks - I want to make sure I understand what substitutions to add (likely that's what I need to do in the lit.config.py)
> > But! picking up from @dyung and @sscalpone's comments: if you are already doing `-DCOMPILER_RT_BUILD_SANITIZERS=OFF` and/or `-DCOMPILER_RT_BUILD_MEMPROF=OFF` then the right solution for you should be what @sscalpone suggested, indeed.
> 
> Thanks for confirming. I can add the define to our build, but would greatly prefer if the build would check whether all of the required components are being built and only then enable it.

I think I found a simple repro: adding `-DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF`. #92679 should address it.

**Comment by mtrofin** [comment]
> > > I'm going to try to find a simpler reproducer.
> > 
> > 
> > Thanks - I want to make sure I understand what substitutions to add (likely that's what I need to do in the lit.config.py)
> 
> Ok, could you try if you can reproduce if you do a monorepo build with:
> 
> ```
> cmake -G Ninja <source>/llvm -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLLVM_CCACHE_BUILD=ON -DLLVM_TARGETS_TO_BUILD=host -DLLVM_ENABLE_PROJECTS='clang;compiler-rt' -DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF
> ```
> 
> (`-DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF` should be the key option, the rest is just my "quick build" defaults)

`ninja check-ctx_profile` works with PR #92679 applied.
